lib/unison-util-relation/src/Unison/Util/Relation.hs

```diff
@@ -500,44 +502,48 @@ compactSet = S.fold (S.union . fromMaybe S.empty) S.empty
 -- | Domain restriction for a relation. Modeled on z.
 (<|), restrictDom :: (Ord a, Ord b) => Set a -> Relation a b -> Relation a b
 restrictDom = (<|)
-s <| r =
-  fromList $
-    concatMap (\(x, y) -> zip (repeat x) (S.toList y)) (M.toList domain')
+s <| r = go s (domain r)
   where
-    domain' = M.unions . List.map filtrar . S.toList $ s
-    filtrar x = M.filterWithKey (\k _ -> k == x) dr
-    dr = domain r -- just to memoize the value
+    go _ Map.Tip = mempty
+    go s _ | Set.null s = mempty
+    go s (Map.Bin _ amid bs l r) = here <> go sl l <> go sr r
+      where
+        (sl, hasMid, sr) = Set.splitMember amid s
+        mids = Set.singleton amid
+        here =
+          if hasMid
+            then Relation (Map.singleton amid bs) (Map.fromList $ (,mids) <$> (Set.toList bs))
```

Comment:
```suggestion
            then Relation (Map.singleton amid bs) (Map.fromAscList $ (,mids) <$> (Set.toList bs))
```

Reply:
Oop missed this. Good idea. I don’t think it would make much difference since these will be smallish sets for at least the relations we are dealing with - you might have some name conflicts, but not like thousands for a given key.